### PR TITLE
Merging last half a year of Development branch

### PR DESCRIPTION
**Pull request by marcinpaluch1994**

### CODE REVIEW

**Comment by marcinpaluch1994** [comment]
Requires new variables in config training:
LR_INITIAL, LR_MINIMAL, LR_PATIENCE, LR_DECREASE_FACTOR - allows for setting parameters of learning rate scheduler from config.

 VALIDATE_ALSO_ON_TRAINING_SET - for proper assessment of overfitting the additional validation on training set after each epoch can be calculated ("training loss" as of default is calculated during the training and before gradient step is taken)
